XMonad/Util/NamedScratchpad.hs

```diff
@@ -218,7 +234,11 @@ scratchpadWorkspaceTag = "NSP"
 -- | Manage hook to use with named scratchpads
 namedScratchpadManageHook :: NamedScratchpads -- ^ Named scratchpads configuration
                           -> ManageHook
-namedScratchpadManageHook = composeAll . fmap (\c -> query c --> hook c)
+namedScratchpadManageHook nsps = do
+    ns <- liftX $ XS.get >>= \case
+        NSPState [] -> nsps <$ XS.put (NSPState nsps) -- initialise
+        NSPState xs -> pure xs
```

Comment:
This is quite a hack, but I couldn't see another way of making this a backwards compatible change :/